python/lsst/ts/watcher/rules/cp_verify_alarm.py

```diff
@@ -0,0 +1,305 @@
+        )  # majority of detectors
+
+        # Define failure threshold per exposure
+        failure_threshold_exposure = (
+            max_number_failures_per_detector_per_test * max_number_failed_detectors
+        )
+
+        # Count the number of failures per test per exposure.
+        total_counter_failed_tests = {}
+        for exposure in [key for key in verify_stats if key != "SUCCESS"]:
+            if "FAILURES" in verify_stats[exposure]:
+                # Gather the names of the types of tests that failed.
+                # 'stg' is of the form e.g., 'R22_S21 C17 NOISE' or
+                # 'R22_S22 SCATTER', so we retreive the name of the test from
+                # the last element after splitting it.
+                fail_count = [
+                    stg.split(" ")[-1] for stg in verify_stats[exposure]["FAILURES"]
+                ]
+                counter = {}
+                for test in fail_count:
+                    if test in counter:
+                        counter[test] += 1
+                    else:
+                        counter[test] = 1
+                total_counter_failed_tests[exposure] = counter
+            else:
+                continue
+
+        # If there are not exposures with tests that failed.
+        if len(total_counter_failed_tests) == 0:
+            return certify_calib, None, None
+
+        # Count the number of exposures where a given test fails
+        # in the majority of detectors.
+        failed_exposures_counter = 0
+        for exposure in total_counter_failed_tests:
+            for test in total_counter_failed_tests[exposure]:
+                if (
+                    total_counter_failed_tests[exposure][test]
+                    >= failure_threshold_exposure
+                ):
+                    failed_exposures_counter += 1
+                    # Exit the inner loop over tests: just need
+                    # the condition to be satisfied for
+                    # at least one type of test
+                    break
+
+        # For at least one type of test, if the majority of tests fail in
+        # the majority of detectors and the majority of exposures,
+        # then don't certify the calibration
+        if failed_exposures_counter >= max_number_failed_exposures:
+            certify_calib = False
+
+        # Return a dictionary with the thresholds to report
+        # them if verification fails.
+        thresholds = {
+            "MAX_FAILURES_PER_DETECTOR_PER_TEST_TYPE_THRESHOLD": max_number_failures_per_detector_per_test,
+            "MAX_FAILED_DETECTORS_THRESHOLD": max_number_failed_detectors,
+            "MAX_FAILED_TESTS_PER_EXPOSURE_THRESHOLD": failure_threshold_exposure,
+            "MAX_FAILED_EXPOSURES_THRESHOLD": max_number_failed_exposures,
+            "FINAL_NUMBER_OF_FAILED_EXPOSURES": failed_exposures_counter,
+        }
+
+        return certify_calib, total_counter_failed_tests, thresholds
```

Comment:
I suggest returning a dataclass with 3 named attributes. Advantages:
* Clarifies the code and makes it more robust, because you reference attributes by name, not position.
* Adds flexibility: you can add more fields later without breaking calling code.
* Simplifies early return if you specify defaults for some items -- you only need to pass the items you actually know to the constructor.

`thresholds` is another good candidate for a dataclass, depending on what you do with it. (And you can trivially turn a dataclass instance into a dict using `vars(my_dataclass_instance)`)

Reply:
I turned the return variables into a dataclass and I see your points. As for the `thresholds` dictionary, it's only used once to get statistics to report in the Alarm, so I'm not sure if there's much difference. I'll be happy to do it in this case too if you still think it's better.

Reply:
I gently suggest it, but it's OK as is.

Reply:
I turned `thresholds` into a dataclass too.